File: src/main/java/com/gmail/samueler53/fwhumanstratego/listeners/onPlayerInventoryClickListener.java
package com.gmail.samueler53.fwhumanstratego.listeners;

import com.gmail.samueler53.fwhumanstratego.FwHumanStratego;
import com.gmail.samueler53.fwhumanstratego.managers.GameManager;
import com.gmail.samueler53.fwhumanstratego.message.Message;
import com.gmail.samueler53.fwhumanstratego.objects.Game;
import com.gmail.samueler53.fwhumanstratego.objects.Squad;
import org.bukkit.Bukkit;
import org.bukkit.GameMode;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryType;

import java.util.Objects;
import java.util.UUID;

public class onPlayerInventoryClickListener implements Listener {

    GameManager gameManager = GameManager.getInstance();

    @EventHandler

    public void OnPlayerInventoryClick(InventoryClickEvent event) {
        if (event.getWhoClicked() instanceof Player) {
            Player player = (Player) event.getWhoClicked();
            UUID uuid = player.getUniqueId();
            if (gameManager.isPlayerPlaying(uuid).isPresent()) {
                Game game = gameManager.getGameWherePlayerPlaying(uuid);
                if (isClickingOnArmor(event.getSlotType())) {
                    Message.GAME_TOGGLEARMOR.send(player);
                    event.setCancelled(true);
                }
                try {
                    if (player.getGameMode().equals(GameMode.SPECTATOR)) {
                        event.setCancelled(true);
                    } else if (game.isStarted()) {
                        Squad squad = game.getSquadFromPlayer(uuid);
                        Material treasure = squad.getTreasure();
                        Squad otherSquad = game.getOtherSquad(squad);
                        Material otherTreasure = otherSquad.getTreasure();
                        if (event.getCurrentItem() != null && isStealingTreasure(event.getCurrentItem().getType(), treasure)) {
                            Message.GAME_STEALHISWOOL.send(player);
                            event.setCancelled(true);
                        } else if (Objects.equals(Objects.requireNonNull(event.getClickedInventory()).getLocation(), game.getArena().getTreasureBlueLocation()) && squad.getName().equalsIgnoreCase("red")) {
                            if (isStealingTreasure(event.getCurrentItem().getType(), otherTreasure)) {
                                Message.GAME_STEALEDWOOLBLUE.broadcast(game, player.getDisplayName());
                            }
                        } else if (Objects.equals(Objects.requireNonNull(event.getClickedInventory()).getLocation(), game.getArena().getTreasureRedLocation()) && squad.getName().equalsIgnoreCase("blue")) {
                            if (isStealingTreasure(event.getCurrentItem().getType(), otherTreasure)) {
                                Message.GAME_STEALEDWOOLRED.broadcast(game, player.getDisplayName());
                            }
                        } else if (event.getView().getTopInventory().getLocation() != null && event.getCurrentItem() != null && (event.getCurrentItem().getType().equals(Material.RED_WOOL) || event.getCurrentItem().getType().equals(Material.BLUE_WOOL)) && (!Objects.equals(event.getView().getTopInventory().getLocation(), game.getArena().getTreasureRedLocation()) && !Objects.equals(event.getView().getTopInventory().getLocation(), game.getArena().getTreasureBlueLocation()))) {
                            event.setCancelled(true);
                        }
                        if ((Objects.equals(event.getInventory().getLocation(), game.getArena().getTreasureRedLocation()) && squad.getName().equalsIgnoreCase("red")) || (Objects.equals(event.getInventory().getLocation(), game.getArena().getTreasureBlueLocation()) && squad.getName().equalsIgnoreCase("blue"))) {
                            Bukkit.getScheduler().runTaskLater(FwHumanStratego.getPlugin(), () -> {
                                if (event.getInventory().contains(otherTreasure)) {
                                    stoleWool(squad, game);
                                }
                            }, 1);
                        }
                    }
                } catch (Exception ignored) {

                }
            }
        }
    }

    private boolean isClickingOnArmor(InventoryType.SlotType slotType) {
        return slotType.equals(InventoryType.SlotType.ARMOR);
    }

    private boolean isStealingTreasure(Material currentItem, Material treasure) {
        return currentItem.equals(treasure);
    }

    private void stoleWool(Squad squad, Game game) {
        Message.GAME_TREASURESTEALED.broadcast(game);
        squad.addPoints(FwHumanStratego.getDefaultConfig().getInt("pointsFromTreasure"));
        game.endRound(squad);
    }
}

File: src/main/java/com/gmail/samueler53/fwhumanstratego/listeners/OnPlayerInventoryClickListener.kt
package com.gmail.samueler53.fwhumanstratego.listeners

import com.gmail.samueler53.fwhumanstratego.FwHumanStratego
import com.gmail.samueler53.fwhumanstratego.managers.GameManager
import com.gmail.samueler53.fwhumanstratego.message.Message
import com.gmail.samueler53.fwhumanstratego.objects.Team
import com.gmail.samueler53.fwhumanstratego.utils.launch
import kotlinx.coroutines.delay
import org.bukkit.GameMode
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryType.SlotType

class OnPlayerInventoryClickListener : Listener {

    @EventHandler
    fun onPlayerInventoryClick(event: InventoryClickEvent) {
        val player = event.whoClicked as? Player ?: return
        val game = GameManager.getGameForPlayer(player) ?: return

        if (event.slotType == SlotType.ARMOR) {
            Message.GAME_TOGGLEARMOR.send(player)
            event.isCancelled = true
        }

        if (player.gameMode == GameMode.SPECTATOR) {
            event.isCancelled = true
            return
        }

        if (!game.isStarted) return
        val team = game.getTeamForPlayer(player)

        with(event) {
            when {
                currentItem != null && currentItem!!.type == team.treasure -> {
                    Message.GAME_STEALHISWOOL.send(player)
                    isCancelled = true
                }
                clickedInventory!!.location == game.arena.treasureBlueLocation &&
                    team.type == Team.Type.RED -> {
                    if (currentItem!!.type == game.getOtherTeam(team).treasure) {
                        Message.GAME_STEALEDWOOLBLUE.broadcast(game, player.displayName)
                    }
                }
                clickedInventory!!.location == game.arena.treasureRedLocation &&
                    team.type == Team.Type.BLUE -> {
                    if (currentItem!!.type == game.getOtherTeam(team).treasure) {
                        Message.GAME_STEALEDWOOLRED.broadcast(game, player.displayName)
                    }
                }
                view.topInventory.location != null && currentItem != null &&
                    (currentItem!!.type == Material.RED_WOOL || currentItem!!.type == Material.BLUE_WOOL) &&
                    view.topInventory.location != game.arena.treasureRedLocation &&
                    view.topInventory.location != game.arena.treasureBlueLocation -> {
                    isCancelled = true
                }
            }

            if (inventory.location == game.arena.treasureRedLocation &&
                team.type == Team.Type.RED || inventory.location == game.arena.treasureBlueLocation &&
                team.type == Team.Type.BLUE
            ) {
                if (inventory.contains(game.getOtherTeam(team).treasure)) {
                    launch {
                        delay(1)
                        Message.GAME_TREASURESTEALED.broadcast(game)
                        team.addPoints(FwHumanStratego.defaultConfig.getInt("pointsFromTreasure"))
                        game.endRound(team)
                    }
                }
            }
        }
    }
}